future/atomicTransactionComposer.go

```diff
@@ -0,0 +1,540 @@
+		}
+
+		return stxs, txids, nil
+	}
+}
+
+/**
+ * Create a TransactionSigner that can sign transactions for the provided Multisig account.
+ * @param msig - The Multisig account metadata
+ * @param sks - An array of private keys belonging to the msig which should sign the transactions.
+ */
+func MakeMultiSigAccountTransactionSigner(msig crypto.MultisigAccount, sks [][]byte) TransactionSigner {
+
+	return func(txGroup []types.Transaction, indexesToSign []int) ([][]byte, []string, error) {
+		var stxs [][]byte
+		var txids []string
+		for _, pos := range indexesToSign {
+			var unmergedStxs [][]byte
+			var txid string
+			for _, sk := range sks {
+				tempTxid, unmergedStxBytes, err := crypto.SignMultisigTransaction(sk, msig, txGroup[pos])
+				txid = tempTxid
+				if err != nil {
+					return nil, nil, err
+				}
+
+				unmergedStxs = append(unmergedStxs, unmergedStxBytes)
+			}
+
+			if len(sks) > 1 {
+				tempTxid, stxBytes, err := crypto.MergeMultisigTransactions(unmergedStxs...)
+				txid = tempTxid
+				if err != nil {
+					return nil, nil, err
+				}
+
+				stxs = append(stxs, stxBytes)
+				txids = append(txids, txid)
+			} else {
+				stxs = append(stxs, unmergedStxs[0])
+				txids = append(txids, txid)
+			}
+		}
+
+		return stxs, txids, nil
+	}
+}
+
+/** Represents an unsigned transactions and a signer that can authorize that transaction. */
+type TransactionWithSigner struct {
+	/** An unsigned transaction */
+	Txn types.Transaction
+	/** A transaction signer that can authorize txn */
+	Signer TransactionSigner
+}
+
+type MethodArgument interface {
+	IsMethodArgument()
+}
+
+type ABIValue abi.Value
+
+func (abiValue ABIValue) IsMethodArgument()                 {}
+func (txAndSigner TransactionWithSigner) IsMethodArgument() {}
+
+type AtomicTransactionComposerStatus = int
+
+const (
+	/** The atomic group is still under construction. */
+	BUILDING AtomicTransactionComposerStatus = iota
+
+	/** The atomic group has been finalized, but not yet signed. */
+	BUILT
+
+	/** The atomic group has been finalized and signed, but not yet submitted to the network. */
+	SIGNED
+
+	/** The atomic gorup has been finalized, signed, and submitted to the network. */
+	SUBMITTED
+
+	/** The atomic group has been finalized, signed, submitted, and successfully committed to a block. */
+	COMMITTED
+)
+
+/** The maximum size of an atomic transaction group. */
+const MAX_GROUP_SIZE = 16
+
+/** A class used to construct and execute atomic transaction groups */
+type AtomicTransactionComposer struct {
+	/** The current status of the composer. The status increases monotonically. */
+	status AtomicTransactionComposerStatus
+
+	/** The transactions in the group with their respective signers. If status is greater then BUILDING
+	then this slice cannot change. */
+	transactions []TransactionWithSigner
+
+	/** Method calls constructed from information passed into AddMethodCall(). Used in Execute() to extract
+	return values. */
+	methodCalls []Method
```

Comment:
+1 Good idea
The `Method`s lining up with `transaction` is nice and we don't have to separately keep track of the `method` txIDs, since we already need to query the logs for each method call in `execute()`.